File: 백준/Silver/13305. 주유소/주유소.go
package main

import (
	"fmt"
)

func max(a int, b int) int {
	if a > b {
		return a
	}
	return b
}

func main() {
	const LMIT = 100001
	const MAX = 1000000000
	var dis [LMIT]int
	var grd [LMIT]int
	mincost := MAX
	N := 0
	result := 0
	mindex := 0
	L := 0
	fmt.Scan(&N)
	for i := 1; i < N; i++ {
		fmt.Scan(&dis[i])
		L += dis[i]
	}
	for i := 1; i < N; i++ {
		fmt.Scan(&grd[i])
		if grd[i] < mincost {
			mincost = grd[i]
			mindex = i
		}
	}
	if mindex == 1 {
		result = grd[1] * L
	} else {
		mincost = grd[1]
		result += mincost * dis[1]
		L -= dis[1]
		for i := 2; i < mindex; i++ {
			if mincost < grd[i] {
				result += mincost * dis[i]
				L -= dis[i]
			} else {
				mincost = grd[i]
				result += mincost * dis[i]
				L -= dis[i]
			}
		}
		result += grd[mindex] * L
	}
	fmt.Print(result)

}

File: 백준/Silver/13305. 주유소/주유소.py
import sys
sys.setrecursionlimit(2500)


if __name__ == "__main__":
    N=int(sys.stdin.readline())
    result=0
    mindex=0
    dis=list(map(int,input().split()))
    L=sum(dis)
    grd=list(map(int,input().split()))
    mincost=min(grd)
    mindex=grd.index(mincost)

    if mindex==0:
        result=L*mincost
    else:
        mincost=grd[0]
        result=dis[0]*mincost
        L-=dis[0]
        for i in range(1,mindex):
            if mincost<grd[i]:
                result+=mincost*dis[i]
                L-=dis[i]
            else:
                mincost=grd[i]
                result+=mincost*dis[i]
                L-=dis[i]
        result+=grd[mindex]*L
    print(result)
   